### ISSUE TITLE
HTML5 Exports cause high CPU usage in Chrome

### ISSUE THREAD

**Issue by MarshallJDem** [godotengine/godot#37473]
**Godot version:**
3.2-Stable

**OS/device including version:**
Windows 10, Mac OSX Catalina 10.15.4, Google Chrome V 80.0.3987.149 

**Issue description:**
When exporting a game for HTML5, if you run it in Google Chrome, it takes a very large chunk of the CPU. This happens indefinitely - most browsers experience a spike when loading up the game, but Chrome maintains this high CPU. I ran it for 5 minutes and it did not drop, whereas Firefox for example drops in CPU usage a few seconds after startup.

The CPU usage for chrome is coming from something called "Software Reporter Tool". This may mean Godot is raising a red flag and chrome is using high CPU to monitor the program, but that's just a guess of mine after some brief research.

This has been tested on two different computers, running the game directly out of Godot as well as off an Apache server.

This is a photo of running the game directly in the editor. There is no abnormal CPU usage.
![2020-03-31 (9)](https://user-images.githubusercontent.com/24983657/78070411-5434f480-7361-11ea-80c8-7768535b648a.png)

This is a photo of running the game using the HTML5 export on chrome. This has high CPU usage.
![2020-03-31 (10)](https://user-images.githubusercontent.com/24983657/78070493-7fb7df00-7361-11ea-81bb-de81138d4a9a.png)

**Steps to reproduce:**
Run any Godot project on HTML5 export on Chrome.
**Minimal reproduction project:**
This can be done with a new project that has nothing other than an empty scene file.

**Comment by Calinou** [comment]
If this is a non-game application (i.e. smooth animations aren't important), try enabling **Application > Run > Low Processor Mode** in the Project Settings. Other than that, I don't think we can do much about the issue at hand.

**Comment by MarshallJDem** [comment]
This is a game, however luckily in my case it is 2D and the CPU usage is low enough anyway that performance hasn't been an issue. I just have noticed that my fan has been turning into a tornado when I'm running the game. 

But just curious - is this unsolvable because you think that chrome is to blame for this?

**Comment by Calinou** [comment]
> But just curious - is this unsolvable because you think that chrome is to blame for this?

Unfortunately, that might be the case. On Windows, Chrome uses ANGLE to render WebGL, which means WebGL is always translated to Direct3D before being rendered. This incurs an additional cost. I presume Google did it this way because they didn't want to rely on low-quality OpenGL drivers, but it's a shame even NVIDIA users have to pay for this cost (even though their OpenGL drivers are good).

**Comment by MarshallJDem** [comment]
I'll drink to that... WebGL does not have the best profiling :P. 

A bit of a good news on this is that having additional tabs of the game open doesn't scale CPU usage linearly, so it seems like Chrome is at least adjusting to higher loads which would imply that bigger projects won't necessarily suffer more than mine.

If this becomes more of an issue I'll dive into this and post any solutions I find in this thread. Thanks for the reply!

**Comment by Giwayume** [comment]
Chrome & other browsers have throttled requestAnimationFrame on inactive tabs for a long time now. In some cases, you can get less than 1fps which obviously saves a lot of CPU.

**Comment by davehayes** [comment]
I have another observation that may relate to this issue. Consider this code, which is found in your html5 export:

```javascript
var animationCallbacks = [];
function animate(time) {
	animationCallbacks.forEach(callback => callback(time));
	requestAnimationFrame(animate);
}
requestAnimationFrame(animate);

function adjustCanvasDimensions() {
	var scale = window.devicePixelRatio || 1;
	var width = window.innerWidth;
	var height = window.innerHeight;
	canvas.width = width * scale;
	canvas.height = height * scale;
	canvas.style.width = width + "px";
	canvas.style.height = height + "px";
}
animationCallbacks.push(adjustCanvasDimensions);
```

Is there really a need to alter the canvas width and height every animation tick?

**Comment by Giwayume** [comment]
Seems like an easy way to guarantee it's always sized correctly, any reason not to do this, though?
```
window.addEventListener('resize', adjustCanvasDimensions);
```

**Comment by Calinou** [comment]
For reference, here are the benchmark results I get using Firefox 77:

![image](https://user-images.githubusercontent.com/180032/85372224-d16fa000-b531-11ea-85d4-e3c820b5593b.png)

**Comment by davehayes** [comment]
> Seems like an easy way to guarantee it's always sized correctly, any reason not to do this, though?

That was my implied question.

Your solution is almost exactly what I did to fix this issue and a related one where I could not get a canvas resize to stick. Finding this illustrated why. Here's what's currently on my page:

```
  function adjustCanvasDimensions() {
    if (godotResizeTrigger) {
      godotResizeTrigger = false;
      var scale = window.devicePixelRatio || 1;
      var width = window.innerWidth - godotWidthAdjust;
      var height = window.innerHeight - godotHeightAdjust;
      canvas.width = width * scale;
      canvas.height = height * scale;
      canvas.style.width = width + "px";
      canvas.style.height = height + "px";
      console.log("GODOT code adjusted canvas dimensions to " + canvas.width + " and " + canvas.height);
    }
  }
```
If someone decides to fix this, I would recommend those `*Adjust` variables above. My use case specifically needs the Godot canvas in the bottom half of the page, and without those variables you cannot achieve this.

**Comment by Faless** [comment]
The `adjustCanvasDimesions` function has been improved in #42178 (and backported via #42266).
The changes will be available in 3.2.4 . Additionally, the `low_processor_usage_mode` has been also fixed via #40018 .
Closing this a fixed (browser version will most likely still be less performant than desktop version, but according to my profiling this specific issue is fixed and I'm registering lower CPU usage).
